### PR TITLE
 feat(scheduling): add zfs pool capacity tracking 

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

**Why is this PR required? What issue does it fix?**:

This PR 
 - add ZFSNode CRD
 - add CapacityWeighted scheduler

**What this PR does?**:

**Does this PR require any upgrade changes?**:

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

File: deploy/yamls/zfsnode-crd.yaml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  names:
+    kind: ZFSNode
+    listKind: ZFSNodeList
+    plural: zfsnodes
+    shortNames:
+    - zfsnode
+    singular: zfsnode
+  scope: Namespaced
+  versions:
+  - name: v1
```
Comment (akhilerm) [inline deploy/yamls/zfsnode-crd.yaml:31]: Is the version `v1` or `v1alpha1`?

File: pkg/apis/openebs.io/zfs/v1/register.go [hunk 2]
```diff
@@ -77,6 +77,8 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 		&ZFSBackupList{},
 		&ZFSRestore{},
 		&ZFSRestoreList{},
+		&ZFSNode{},
```
Comment (akhilerm) [inline pkg/apis/openebs.io/zfs/v1/register.go:80]: @pawanpraka1 Doesn't this CR need to be in `v1alpha1` or all ZFS CRs are `v1`

File: pkg/apis/openebs.io/zfs/v1/zfsnode.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced,shortName=zfsnode
+type ZFSNode struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Pools []Pool `json:"pools"`
+}
+
+// Pool specifies attributes of a given zfs pool exists on node.
```
Comment (akhilerm) [inline pkg/apis/openebs.io/zfs/v1/zfsnode.go:41]:
```suggestion
// Pool specifies attributes of a given zfs pool that exists on the node.
```

File: pkg/driver/agent.go [hunk 4]
```diff
@@ -54,6 +55,14 @@ func NewNode(d *CSIDriver) csi.NodeServer {
 	// set up signals so we handle the first shutdown signal gracefully
 	stopCh := signals.SetupSignalHandler()
 
+	// start the zfsnode resource watcher
+	go func() {
+		err := zfsnode.Start(&ControllerMutex, stopCh)
+		if err != nil {
+			klog.Fatalf("Failed to start LVM node controller: %s", err.Error())
```
Comment (akhilerm) [inline pkg/driver/agent.go:62]:
```suggestion
			klog.Fatalf("Failed to start ZFS node controller: %s", err.Error())
```

File: pkg/driver/controller.go [hunk 5]
```diff
@@ -784,15 +849,97 @@ func (cs *controller) ControllerPublishVolume(
...
+		segments = topology.Segments
+	}
+	nodeNames, err := cs.filterNodesByTopology(segments)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	zfsNodesCache := cs.zfsNodeInformer.GetIndexer()
+	params := req.GetParameters()
+	vgParam := helpers.GetInsensitiveParameter(&params, "volgroup")
```
Comment (akhilerm) [inline pkg/driver/controller.go:871]: `s/volgroup/poolname/`.

volgroup needs to be changed to pool name

File: pkg/driver/controller.go [hunk 6]
```diff
@@ -784,15 +849,97 @@ func (cs *controller) ControllerPublishVolume(
...
+		if err != nil {
+			klog.Warning("unexpected error after querying the zfsNode informer cache")
+			continue
+		}
+		if !exists {
+			continue
+		}
+		zfsNode := v.(*zfsapi.ZFSNode)
+		// rather than summing all free capacity, we are calculating maximum
+		// lv size that gets fit in given vg.
```
Comment (akhilerm) [inline pkg/driver/controller.go:885]: Change comment as per ZFS terms.

File: pkg/driver/controller.go [hunk 7]
```diff
@@ -784,15 +849,97 @@ func (cs *controller) ControllerPublishVolume(
...
+		// See https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1472-storage-capacity-tracking#available-capacity-vs-maximum-volume-size &
+		// https://github.com/container-storage-interface/spec/issues/432 for more details
+		for _, vg := range zfsNode.Pools {
```
Comment (akhilerm) [inline pkg/driver/controller.go:888]:
```suggestion
		for _, zpool := range zfsNode.Pools {
```

File: pkg/mgmt/zfsnode/zfsnode.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+	"github.com/openebs/zfs-localpv/pkg/zfs"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+func (c *NodeController) listZFSVolumeGroup() ([]apis.Pool, error) {
```
Comment (akhilerm) [inline pkg/mgmt/zfsnode/zfsnode.go:36]:
```suggestion
func (c *NodeController) listZFSPool() ([]apis.Pool, error) {
```

File: pkg/mgmt/zfsnode/zfsnode.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	cachedNode, err := c.NodeLister.ZFSNodes(namespace).Get(name)
+	if err != nil && !k8serror.IsNotFound(err) {
+		return err
+	}
+	var node *apis.ZFSNode
+	if cachedNode != nil {
+		node = cachedNode.DeepCopy()
+	}
+
+	pools, err := c.listZFSVolumeGroup()
```
Comment (akhilerm) [inline pkg/mgmt/zfsnode/zfsnode.go:66]:
```suggestion
	pools, err := c.listZFSPool()
```

**Review by akhilerm — CHANGES_REQUESTED** [review]
given some comments.

**Review by akhilerm — APPROVED** [review]
LGTM

File: deploy/yamls/zfsnode-crd.yaml (hunk 1 above)
Comment (akhilerm) [reply deploy/yamls/zfsnode-crd.yaml:31]: We will be using the v1 version for zfsnode CRs

File: pkg/mgmt/zfsnode/zfsnode.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+		// Node resource to be synced.
+		if err := c.syncHandler(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		klog.Infof("Successfully synced '%s'", key)
```
Comment (pawanpraka1) [inline pkg/mgmt/zfsnode/zfsnode.go:274]: let us remove this (or put it uder more verbose) as it will pollute the log every minute.

**Review by pawanpraka1 — APPROVED** [review]
looks good.

**Pull request merged**